runtime/src/chpl-qsbr.c

```diff
@@ -0,0 +1,226 @@
+static void waitForReaders(uint64_t targetEpoch, struct tls_node *list, struct defer_data ddata);
+static void waitForReaders(uint64_t targetEpoch, struct tls_node *list, struct defer_data ddata) {
+  bool canDelete = true;
+  for (struct tls_node *node = list; canDelete && node != NULL; node = node->next) {
+    // Look for a thread which is active but is not on the current global epoch and defer
+    // deletion to them.
+    if (node->nTasks && node->epoch < targetEpoch) {
+        canDelete = false;
+
+        // Create new defer node that contains both the target epoch and data.
+        struct defer_node *defer = chpl_mem_calloc(1, sizeof(*defer), 0, 0, 0);
+        defer->targetEpoch = targetEpoch;
+        defer->ddata = ddata;
+
+        // Push...
+        struct defer_node *old_head;
+        do {
+            old_head = (struct defer_node *) atomic_load_uintptr_t((uintptr_t *) &node->deferList);
+            defer->next = old_head;
+        } while (!atomic_compare_exchange_weak_uintptr_t((uintptr_t *) &node->deferList, (uintptr_t) old_head, (uintptr_t) defer));
+    }
+  }
+
+  if (canDelete) {
+    // Treat as an array?
+    if (ddata.numData) {
+        void **arr = ddata.data;
+        for (int i = 0; i < ddata.numData; i++) {
+            chpl_mem_free(arr[i], 0, 0);
+        }
+    }
+
+    chpl_mem_free(ddata.data, 0, 0);
+  }
+}
+
+void chpl_qsbr_init(void) {
+    CHPL_TLS_INIT(chpl_qsbr_readerTLS);
+}
+
+void chpl_qsbr_onTaskCreation(void) {
+  struct tls_node *node = CHPL_TLS_GET(chpl_qsbr_readerTLS);
+  if (node == NULL) {
+    init_tls();
+    node = CHPL_TLS_GET(chpl_qsbr_readerTLS);
+  }
+
+  // Notify that we are now in-use.
+  node->nTasks++;
+}
+
+
+void chpl_qsbr_onTaskDestruction(void) {
+  struct tls_node *tls = CHPL_TLS_GET(chpl_qsbr_readerTLS);
+  if (tls == NULL) {
+    init_tls();
+    tls = CHPL_TLS_GET(chpl_qsbr_readerTLS);
+  }
+
+  // Notify that we have one less task
+  tls->nTasks--;
+
+  // If we are the last task, empty our defer list...
+  while (tls->nTasks == 0 && tls->deferList) {
+    // Pop
+    struct defer_node *dnode;
+    do {
+      dnode = (struct defer_node *) atomic_load_uintptr_t((uintptr_t *) &tls->deferList);
+    } while (!atomic_compare_exchange_weak_uintptr_t((uintptr_t *) &tls->deferList, (uintptr_t) dnode, (uintptr_t) dnode->next));
+
+    // Only process threads that come after us as a previous thread would have been processed
+    // before us and newly registered threads are pushed as the head of the tls_list.
+    waitForReaders(dnode->targetEpoch, tls->next, dnode->ddata);
+    chpl_mem_free(dnode, 0, 0);
+  }
+}
+
+void chpl_qsbr_checkpoint(void) {
+    struct tls_node *tls = CHPL_TLS_GET(chpl_qsbr_readerTLS);
+    if (tls == NULL) {
+        init_tls();
+        tls = CHPL_TLS_GET(chpl_qsbr_readerTLS);
+    }
+
+    // Observe the current epoch.
+    tls->epoch = global_epoch;
+
+    // Handle a single deferred object we have, if any
+    if (tls->deferList) {
```

Comment:
I think this read needs to be atomic

Reply:
Is there an easier way of handling an atomic read for an object?

Reply:
You could make a helper fn but I don't know of one.